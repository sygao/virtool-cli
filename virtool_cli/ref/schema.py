from pydantic import BaseModel, computed_field

from virtool_cli.utils.models import Molecule


class Segment(BaseModel):
    """The metadata of the segment"""

    name: str
    """The name of the segment"""

    required: bool
    """Whether this segment must be present in all additions."""

    length: int | None = None


class OTUSchema(BaseModel):
    """A schema for the intended data"""

    molecule: Molecule
    """The molecular metadata for this OTU."""

    segments: list[Segment]
    """The segments contained in this OTU."""

    @computed_field
    def multipartite(self) -> bool:
        return len(self.segments) > 1
