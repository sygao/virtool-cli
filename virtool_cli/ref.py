import sys
from pathlib import Path
import click
import structlog

import virtool_cli.build
import virtool_cli.divide
import virtool_cli.isolate
import virtool_cli.repair
import virtool_cli.taxid
from virtool_cli.migrate import run as run_migrate
from virtool_cli.update import run as run_update
from virtool_cli.accessions.catalog import run as run_catalog

ERROR_MESSAGE = click.style("ERROR: ", fg='red')

logger = structlog.get_logger()
    
@click.group("ref")
def ref():
    """
    Commands related to reference files.
    """
    pass


@ref.command()
@click.option(
    "-src",
    "--src_path",
    required=True,
    type=str,
    help="the path to a database reference directory",
)
@click.option(
    "-o",
    "--output",
    type=str,
    default="reference.json",
    help="the output path for a reference.json file",
)
@click.option("-i", "--indent", is_flag=True)
@click.option(
    "-V",
    "--version",
    default=None,
    type=str,
    help="the version string to include in the reference.json file",
)
def build(src_path, output, indent, version):
    """Build a Virtool reference JSON file from a source directory."""
    build_logger = logger.bind(command='build', src_path=src_path, out_path=output)

    if not Path(src_path).exists():
        build_logger.error('Directory not found at src_path')
        return
    
    try:
        virtool_cli.build.run(Path(src_path), Path(output), indent, version)
    except (FileNotFoundError, NotADirectoryError):
        build_logger.exception('Source directory has critical errors')
    except:
        build_logger.exception('Unexpected exception')


@ref.command()
@click.option(
    "-src",
    "--src_path",
    required=True,
    type=str,
    help="the path to a input reference.json file",
)
@click.option(
    "-o",
    "--output",
    default="src",
    type=str,
    help="the output path for a divided reference directory tree",
)
def divide(src_path, output):
    """Divide a reference.json file from Virtool into a reference directory tree."""
    try:
        if not src_path.endswith(".json"):
            raise TypeError
        virtool_cli.divide.run(Path(src_path), Path(output))
    except (TypeError, FileNotFoundError) as e:
        click.echo(
            ERROR_MESSAGE + "Specified reference file either does not exist or is not a proper JSON file"
        )
        click.echo(e)


@ref.command()
@click.option(
    "-src",
    "--src_path",
    required=True,
    type=str,
    help="the path to the input reference.json file",
)
@click.option("-f", "--force_update", is_flag=True)
def taxid(src_path, force_update):
    """Fetch taxid for all OTU in given reference directory."""
    try:
        click.echo("Scanning OTUs for taxon IDs...")
        virtool_cli.taxid.run(Path(src_path), force_update)
    except (FileNotFoundError, NotADirectoryError) as e:
        click.echo(ERROR_MESSAGE + "Not a valid reference directory")
        click.echo(e)


@ref.command()
@click.option(
    "-src",
    "--src_path",
    required=True,
    type=str,
    help="the path to a reference directory",
)
def isolate(src_path):
    """Fetch new isolates for all OTU in a given reference directory."""
    try:
        virtool_cli.isolate.run(Path(src_path))
    except (FileNotFoundError, NotADirectoryError) as e:
        click.echo("Not a valid reference directory")
        click.echo(e)


@ref.command()
@click.option(
    "-src",
    "--src_path",
    required=True,
    type=str,
    help="the path to a reference directory",
)
def repair(src_path):
    """Fix every OTU in a given reference directory."""
    if not Path(src_path).exists():
        logger.critical('Source directory does not exist')

    try:
        virtool_cli.repair.run(Path(src_path))
    except (FileNotFoundError, NotADirectoryError) as e:
        click.echo(ERROR_MESSAGE + "Not a valid reference directory")
        click.echo(e)


@ref.command()
@click.option(
    "-src",
    "--src_path",
    required=True,
    type=str,
    help="the path to a reference directory",
)
@click.option('--debug/--no-debug', default=False)
def migrate(src_path, debug):
    """Convert a reference directory from v1.x to v2.x"""
    if not Path(src_path).exists():
        logger.critical('Source directory does not exist')

    try:
        run_migrate(Path(src_path))
    except (FileNotFoundError, NotADirectoryError) as e:
        click.echo(ERROR_MESSAGE + "Not a valid reference directory")
        click.echo(e)


@ref.command()
@click.option(
    "-src",
    "--src_path",
    required=True,
    type=str,
    help="the path to a reference directory",
)
@click.option(
    "-cat",
    "--catalog_path",
    required=True,
    type=str,
    default='.cache/catalog',
    help="the path to a catalog directory",
)
@click.option('--debug/--no-debug', default=False)
def catalog(src_path, catalog_path, debug):
    """Update or generate a catalog of all included accessions in a src directory"""
    if not Path(src_path).exists():
        click.echo(ERROR_MESSAGE + 'Source directory does not exist')
        return
    
    catalog_dir = Path(catalog_path)
    if not catalog_dir.exists():
        catalog_dir.mkdir(parents=True)

    try:
        run_catalog(
            src_path=Path(src_path),
            catalog_path=Path(catalog_path),
            debugging=debug
        )
    except (FileNotFoundError, NotADirectoryError) as e:
        click.echo(ERROR_MESSAGE + "Not a valid reference directory")
        logger.exception(e)


@ref.command()
@click.option(
    "-src", "--src_path",
    required=True,
    type=str,
    help="the path to a reference directory",
)
@click.option(
    "-cat",
    "--catalog_path",
    required=True,
    type=str,
    default='.cache/catalog',
    help="the path to a catalog directory",
)
@click.option('--debug/--no-debug', default=False)
def update(src_path, catalog_path, debug):
    """Fetch new isolates for all OTU in a given reference directory."""
    if not Path(catalog_path).exists():
        click.echo("Not a valid catalog directory")
        return

    try:
        run_update(Path(src_path), Path(catalog_path), debugging=debug)
    except (FileNotFoundError, NotADirectoryError) as e:
        click.echo("Not a valid reference directory")
        click.echo(e)

if __name__ == "__main__":
    ref()
