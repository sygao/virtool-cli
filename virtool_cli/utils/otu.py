import json
from pathlib import Path

import aiofiles

from virtool_cli.utils.reference import get_isolate_paths, get_sequence_paths


async def get_otu_accessions(otu_path: Path) -> list:
    """
    Gets all accessions from an OTU directory and returns a list

    :param otu_path: Path to an OTU directory under a reference directory
    :return: A list of all accessions under an OTU
    """
    accession_list = []

    for isolate_path in get_isolate_paths(otu_path):
        for sequence_path in get_sequence_paths(isolate_path):
            with open(sequence_path, "r") as f:
                sequence = json.load(f)

            accession_list.append(
                sequence.get('accession')
            )

    return accession_list


async def label_isolates(otu_path: Path) -> dict:
    """
    Return all isolates present in an OTU directory

    :param otu_path: Path to an OTU directory
    :return: A dictionary of isolates indexed by source_name
    """
    if not otu_path.exists():
        raise FileNotFoundError

    isolates = {}

    for iso_path in get_isolate_paths(otu_path):
        async with aiofiles.open(iso_path / "isolate.json", "r") as f:
            contents = await f.read()
            isolate = json.loads(contents)
        isolates[isolate.get("source_name")] = isolate

    return isolates


async def get_otu_accessions_metadata(otu_path: Path) -> dict:
    """
    Returns sequence metadata for all sequences present under an OTU

    :param otu_path: Path to an OTU directory under a reference directory
    :return: An accession-keyed dict containing all constituent sequence metadata
    """
    # get length and segment metadata from sequences
    all_metadata = {}

    for isolate_path in get_isolate_paths(otu_path):
        for sequence_path in get_sequence_paths(isolate_path):
            sequence_metadata = await get_sequence_metadata(sequence_path)

            accession = sequence_metadata["accession"]

            all_metadata[accession] = sequence_metadata

    return all_metadata


async def read_otu(path: Path) -> dict:
    """
    Returns a json file in dict form

    :param path: Path to an OTU directory under a reference source
    :return: Deserialized OTU data in dict form
    """
    async with aiofiles.open(path / "otu.json", "r") as f:
        contents = await f.read()
        otu = json.loads(contents)

    return otu


async def fetch_exclusions(otu_path: Path) -> list:
    async with aiofiles.open(otu_path / "exclusions.json", "r") as f:
        contents = await f.read()
        exclusions = json.loads(contents)

    return exclusions


async def get_sequence_metadata(sequence_path: Path) -> dict:
    """
    Gets the accession length and segment name from a sequence file
    and returns it in a dict

    :param sequence_path: Path to a sequence file
    :return: A dict containing the sequence accession, sequence length and segment name if present
    """
    sequence = await parse_sequence(sequence_path)

    sequence_metadata = {
        "accession": sequence["accession"],
        "length": len(sequence["sequence"]),
    }

    segment = sequence.get("segment", None)
    if segment is not None:
        sequence_metadata["segment"] = segment

    return sequence_metadata


async def parse_sequence(path):
    async with aiofiles.open(path, "r") as f:
        contents = await f.read()
        sequence = json.loads(contents)

    return sequence
