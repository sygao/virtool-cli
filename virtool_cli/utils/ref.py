import json
import re
from pathlib import Path


def get_otu_paths(src_path: Path) -> list:
    """
    Generates a list of paths to all OTUs in a src directory.

    :param src_path: Path to a src database directory
    :return: List of paths to all OTU in a src directory
    """
    return [otu for otu in src_path.iterdir() if otu.is_dir()]

def generate_otu_dirname(otu_name: str):
    """
    """
    no_whitespace = otu_name.replace(" ", "_")
    no_plus = no_whitespace.replace('+', 'and')
    no_symbols = re.split(r'[()/-]+', no_plus)

    return no_symbols

def parse_otu(path: Path) -> dict:
    """
    Returns a json file in dict form

    :param paths: Path to an OTU in a reference
    :return: OTU data in dict form
    """
    with open(path / "otu.json", "r") as f:
        otu = json.load(f)

    return otu