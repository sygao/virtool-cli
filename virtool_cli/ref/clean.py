import shutil
from pathlib import Path
import structlog
import click

from virtool_cli.utils.logging import configure_logger

base_logger = structlog.get_logger()


def run(repo_path: Path, debugging: bool = False):
    click.echo("Cleaning cache files...")
    cache_path = repo_path / ".cache"
    clean_cache(cache_path)


def clean_cache(cache_path):
    if cache_path.exists():
        shutil.rmtree(cache_path / 'updates')
