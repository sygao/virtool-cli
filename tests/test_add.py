import os
import subprocess
from pathlib import Path

import pytest
from syrupy import SnapshotAssertion
from syrupy.filters import props

from virtool_cli.ref.otu import create_otu_with_schema
from virtool_cli.ref.otu import create_otu, update_otu
from virtool_cli.ref.repo import EventSourcedRepo


VIRTOOL_REF = ["virtool", "ref"]


def run_create_otu_command(
    path: Path, taxid: int, accessions: list | None = None, autofill: bool = False
):
    if accessions is None:
        accessions = []

    options = ["--path", str(path)]
    if autofill:
        options.append("--autofill")

    subprocess.run(
        VIRTOOL_REF + ["otu", "create"] + [str(taxid)] + accessions + options,
        check=False,
    )


def run_update_otu_command(taxid: int, path: Path):
    subprocess.run(
        VIRTOOL_REF + ["otu", "update"] + [str(taxid)] + ["--path", str(path)],
        check=False,
    )


def run_add_sequences_command(taxid: int, accessions: list[str], path: Path):
    subprocess.run(
        VIRTOOL_REF
        + ["sequences", "add"]
        + accessions
        + ["--taxid", str(taxid)]
        + ["--path", str(path)],
        check=False,
    )


class TestCreateOTU:
    def test_empty_success(
        self,
        precached_repo: EventSourcedRepo,
        snapshot: SnapshotAssertion,
    ):
        """Test that an OTU can be created in an empty repository."""
        print(os.listdir(precached_repo.path / "src"))

        otu = create_otu(precached_repo, 345184)

        assert otu.dict() == snapshot(exclude=props("id", "isolates"))

        # Ensure only one OTU is present in the repository, and it matches the return
        # value of the creation function.
        assert precached_repo.get_all_otus() == [otu]

    def test_empty_fail(self, scratch_repo: EventSourcedRepo):
        with pytest.raises(ValueError):
            create_otu(scratch_repo, 345184)

    @pytest.mark.parametrize(
        "taxid, accessions",
        [(1278205, ["NC_020160"]), (345184, ["DQ178610", "DQ178611"])],
    )
    def test_otu_autoschema(
        self,
        taxid: int,
        accessions: list[str],
        precached_repo,
        snapshot: SnapshotAssertion,
    ):
        assert not precached_repo.get_all_otus()

        otu = create_otu_with_schema(
            repo=precached_repo, taxid=taxid, accessions=accessions
        )

        assert precached_repo.get_all_otus()

        assert otu.schema is not None

        assert otu.dict() == snapshot(exclude=props("id"))


class TestCreateOTUCommands:
    @pytest.mark.parametrize(
        "taxid, accessions",
        [(1278205, ["NC_020160"]), (345184, ["DQ178610", "DQ178611"])],
    )
    def test_autoschema(
        self,
        taxid: int,
        accessions: list[str],
        precached_repo: EventSourcedRepo,
        snapshot: SnapshotAssertion,
    ):
        run_create_otu_command(
            taxid=taxid,
            path=precached_repo.path,
            accessions=accessions,
        )

        otus = EventSourcedRepo(precached_repo.path).get_all_otus()

        assert len(otus) == 1
        otu = otus[0]

        assert otu.dict() == snapshot(exclude=props("id", "isolates"))

    @pytest.mark.ncbi()
    def test_autofill(
        self,
        precached_repo: EventSourcedRepo,
        snapshot: SnapshotAssertion,
    ):
        run_create_otu_command(
            taxid=345184,
            path=precached_repo.path,
            accessions=["DQ178610", "DQ178611"],
            autofill=True,
        )

        otus = EventSourcedRepo(precached_repo.path).get_all_otus()

        assert len(otus) == 1
        otu = otus[0]

        assert otu.dict() == snapshot(exclude=props("id", "isolates"))
        assert otu.accessions

    @pytest.mark.ncbi()
    def test_autofill_with_schema(
        self,
        precached_repo: EventSourcedRepo,
        snapshot: SnapshotAssertion,
    ):
        run_create_otu_command(
            taxid=345184,
            accessions=["DQ178610", "DQ178611"],
            path=precached_repo.path,
            autofill=True,
        )

        otus = EventSourcedRepo(precached_repo.path).get_all_otus()

        assert len(otus) == 1
        otu = otus[0]

        assert otu.schema == snapshot
        assert {"DQ178610", "DQ178611"}.intersection(otu.accessions)


class TestAddSequences:
    def test_success(
        self,
        precached_repo: EventSourcedRepo,
        snapshot: SnapshotAssertion,
    ):
        accessions = ["DQ178614", "DQ178613", "DQ178610", "DQ178611"]
        run_add_sequences_command(345184, accessions, precached_repo.path)

        for otu in precached_repo.get_all_otus():
            assert otu.accessions == set(accessions)

            assert otu.dict() == snapshot(exclude=props("id", "isolates"))

            for isolate in otu.isolates:
                assert isolate.dict() == snapshot(exclude=props("id", "sequences"))

                for accession in sorted(isolate.accessions):
                    assert isolate.get_sequence_by_accession(
                        accession,
                    ).dict() == snapshot(exclude=props("id"))


@pytest.mark.ncbi()
class TestUpdateOTU:
    def test_success_no_exclusions(
        self,
        precached_repo: EventSourcedRepo,
        snapshot: SnapshotAssertion,
    ):
        otu = create_otu(precached_repo, 345184)
        update_otu(precached_repo, otu)

        otu = precached_repo.get_otu(otu.id)

        assert [otu.dict() for otu in precached_repo.iter_otus()] == snapshot(
            exclude=props("id", "isolates"),
        )

        assert otu.accessions == {
            "DQ178608",
            "DQ178609",
            "DQ178610",
            "DQ178611",
            "DQ178612",
            "DQ178613",
            "DQ178614",
            "NC_038792",
            "NC_038793",
        }
