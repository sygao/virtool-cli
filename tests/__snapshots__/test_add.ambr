# serializer version: 1
# name: TestCreateOTU.test_empty_success
  dict({
    'acronym': '',
    'excluded_accessions': list([
    ]),
    'legacy_id': None,
    'name': 'Cabbage leaf curl Jamaica virus',
    'schema': None,
    'taxid': 345184,
  })
# ---
# name: TestCreateOTU.test_otu_autoschema[1278205-accessions0]
  dict({
    'acronym': '',
    'excluded_accessions': list([
    ]),
    'isolates': list([
      dict({
        'legacy_id': None,
        'name': dict({
          'type': isolate,
          'value': '4706174',
        }),
        'sequences': list([
          dict({
            'accession': 'NC_020160',
            'definition': 'Dahlia latent viroid, complete sequence',
            'legacy_id': None,
            'segment': '',
            'sequence': 'CAGGTCTTCTAAGGGTTCCTGTGGTGCCTCCCTCCAAGGCCGCGTAGGGAAAGAAAAAGTGAAAGAAAGTGTACCTGAAGAGAGAGAGACTCTCTGAGGAGCCCCGGGGCAACTCCGAGAGTGCTGCGGCAGGAGGAGCCGGGGGCGGAGGTTGTCTCGCTCTTGAGAGCTCCGCTCCTTGTAGCTTTGAGACTACCGCCCTTTTGCTTCCTTCTCGCTGGCTGACTCGAGGACGCGACCGGTGGTACACCAAGAGGTCTCCACCTCCTAGGGTACTTTTTTCTAACACCGATTCCGTACCGAGCGCCGGAGAGTGAAGCGCCCAGGGCTCCAAAGAAGCCC',
          }),
        ]),
      }),
    ]),
    'legacy_id': None,
    'name': 'Dahlia latent viroid',
    'schema': dict({
      'molecule': dict({
        'strandedness': 'single',
        'topology': 'linear',
        'type': 'RNA',
      }),
      'multipartite': False,
      'segments': list([
        dict({
          'length': 342,
          'name': 'Dahlia latent viroid',
          'required': True,
        }),
      ]),
    }),
    'taxid': 1278205,
  })
# ---
# name: TestCreateOTU.test_otu_autoschema[345184-accessions1]
  dict({
    'acronym': '',
    'excluded_accessions': list([
    ]),
    'isolates': list([
      dict({
        'legacy_id': None,
        'name': dict({
          'type': isolate,
          'value': 'CUc-3',
        }),
        'sequences': list([
          dict({
            'accession': 'DQ178610',
            'definition': 'Cabbage leaf curl Jamaica virus clone 1910 Cu3-A8 segment DNA-A, complete sequence',
            'legacy_id': None,
            'segment': 'DNA-A',
            'sequence': 'ACCGGATGGCCGCGATTTTTTGGAGTCATGCGTGGGACCACGCATTAAATGAAATCTAACCAATCACATGCCGCCTGAGAAGGTTAGATATTATGGCGTACTTAGTGCGCTGTGGGCCTATAAAAAGAAATGAAATTTCATTTTATGCTTTACTTCGAAATGCCTAAGCGGGATGCCCCGTGGCGTTCAATGGCGGGGACCTCTAAGGTGTCCCGCAATGCTAACTACTCACCTCGTGCAGGTATAACCCATAAATTTGATAAGGCCGCTGCCTGGGTTAACAGGCCCATGTACAGGAAGCCCAGGATTTATAGGATGTTTAGAAGCCCTGATGTTCCCAGAGGCTGTGAAGGCCCATGTAAGGTCCAATCTTATGAGCAGCGTCATGATATCTCACATGTTGGGAAGGTCATCTGCATTTCGGATATAACACGTGGCAATGGTATTACCCATCGTGTTGGGAAGCGTTTCTGCGTCAAGTCCGTCTACATTTTGGGCAAGATATGGATGGACGAGAATATCAAGCTCAAGAACCACACCAACAGTGTCATGTCTTGGTTGGTTAGGGACAGGAGACCTTATGGCACTCCTATGGAATTTGGGCAAGTGTTCAACATGTTCGACAACGAGCCCAGTACTGCTACTGTGAAGAACGATCTGCGTGATCGTTTTCAAGTCATGCATAAGTTCTACGCCAAGGTAACCGGTGGGCAGTATGCGAGTAACGAGCAGGCGCTGGTGAAGCGGTTCTGGAAGGTCAATAACTACGTGGTGTATAACCATCAAGAAGCAGGGAAATACGAGAATCATACGGAGAACGCTCTGTTATTGTATATGGCATGTACTCATGCCTCTAATCCTGTGTATGCGACATTGAAAATTCGGATCTATTTTTATGATTCGATAACAAATTAATAAATTTTGATTTTTATATCATGATTTTCGTGTACATGAGTTACATAGGATCTGTCCGTTGCGAAACGAACAGCTCTAATTACATGATTGATACCAATAACCCCTAGGTTATCTAAATGAGACATTACAAGGCATTTGAATCTACTTAAATATGTCGGCCCAGAAGCTCTCATCGAACTCGTCCAGACCTGGAAATTGAAGTAGGCTTTGTGGAGATCCAATGCTCTCCGTAGGTTGTGGTTGAACCGGACTTGGATGTGGTAGATCCTGGTTCTGGTGTACAACGGGTCCTCTACGCGTTGTATCTTGAAATACAGGGGATTTGGAACCTCCCAGATAAAAACGGAATTCTCTGCCTGAGCTACAGTGATGCTCTCCCCGGTGCGTGAATCCATTATCAGCGCAATTGATATGGAGGAAGATAGAACACCCGCAGTTCAAATCAATGCGTCTTCTGCGTACAGCTCTCCTTTTAGCAGTCTTGTGCTGTGCTTTGATAGAGGGGGGCTTCAAGGGTGATGAATTTCGCATTTTTGATAGTCCACGCTCTGAGTGATGCGTTTTCCTCTTTGTTGAGGAAACTTATATAACCGCTCCCCTCTCCTGGATTGCACAGCACGATTGAGGGTATGCCACCTTTAATTTGAACTGGTTTGCCGTACTTACAGTTTGATTGCCAGTCCCTTTGGGCCCCAATGAGCTCTTTCCAGTGCTTTAGCTTTAGATATTGCGGAGCTATGTCATCAATGACGTTATATTCCACATTATTTGAAAAGACCTTTGAATTAAAGTCGAGGTGCCCACTCAAATAATTATGTGGTCCTAATGCACGCGCCCACATGGTCTTGCCTGTTCGTGAATCACCTTCCACTATGATACTAATAGGTCTTTCCGGCCGCGCAGCGGCACTCCGACCAAAATAGTCATCTGCCCACTCTTGCATTTCGTCCGGGACGTTAGTAAAGGAGGAGAGGGGAAACGGAGGAGACCAAGGATCCGGAGCCTTACTGAAAATCCTATCGAGGTTACTGGATAGGTTGTGATACTGAAAAAGAAACTTCTCCGGGAGCTTTTCTTTAATGATTTTCATGGCCTCCTCCTTTGTTCCAGAGTTTAGTGCCTCGGCGGCTGCGTCGTTAGCTGTTTGCTGACCGCCTCTAGCACTTCTTCCGTCGATCTGGAACACTCCCCATTCAACTGTATCTCCGTCCTTGTCGATGTAGGACTTGACGTCGGAGCTGGATTTAGCTCCCTGTATGTTCGGATGGAAATGTGCTGACCGGGATGGGGAGACCAAGTCGAAGAATCTGTTATTCGTGCAGTTGTACTTCCCTTCGAACTGTATGAGAACGTGGATATGAGGCTCCCCATTTTCATGAAGCTCTCTGCATATTTTGATGAACTTCTTGTTCACCGGAGTGTTTATTTTTTGTAATTGGGAAAGTGCCTCTTCTTTGCTAAGAGAGCATCGTGGATACGTAAGGAAATAGTTTTTGGCTTTAACTGAGAAAGAGCCCTTTCGTGGCATATTTGTAAATATGGGTGTTCCCCCGATAGCTCTCTCGCTCAAAACTCCTATGAATTGGGGGAACTGGGGGAACTTATATAGTAGAAGTTCCTAAAGGCAAATCAACACGTGGCGGCCATCCGTTATAATATT',
          }),
          dict({
            'accession': 'DQ178611',
            'definition': 'Cabbage leaf curl Jamaica virus clone 1966 Cu3-B8 segment DNA-B, complete sequence',
            'legacy_id': None,
            'segment': 'DNA-B',
            'sequence': 'ACCGGATGGCCGCAAATTTTTTGGCGTCCTGGTGCTGGGGACCATCTCCCGGCCCCCCTTGGCGCACTATCTCTCATCTTTCGACGTGGCGCGTCGATGAGCGTTGGCTAACGCTTATCTCGGTGTTAGACCTTTAATTTGAATTTCGAAATAATTGGTCGTATATATCTATCTTGACCGGTCTTTGTGACGACAAAGCTTCTGACACATTGTACGATATATTGAGCGTGGTCCAATTATATTTTCTCTGCGGAGTTAGACTATCGCTTATTGTTATTCATCTTGTCTATATAATGGACGATAGGTTAATAATTTAAATCGTCTTTGACAATAACATTTAAATAACAAATCATTTTGTAGAGGAGCTGAGAATAATGTATCCTACAAAGTTTAGGCGTGGGACTTCTTACTCTCATAGACGATTTGTTTCACGTAATCCCTCTTTTAAGCGTGGAACTTTGGTTAGACGCACTGATGGGAAACGTCGTAAAGGCCCATCAAGCAAAGCCCATGATGAGCCTAAAATGAAATTACAACGCATACATGAAAATCAATATGGGCCTGAATTTGTCATGACCCACAACTCAGCTCTTTCAACATTCATTAATTTCCCGGTACTTGGGAAGATCGAACCTAACCGAAGCAGGTCGTATATTAAGTTGAATCGTTTATTATTTAAGGGAACCGTTAAGATTGAGCGTGTGCACGCTGATGTGAACATGGACGGAGTAAGTTCGAAGATAGAAGGTGTGTTCTCTCTTGTTATTGTTATGGATCGCAAGCCACATTTGAGTTCCGCTGGAGGTTTGCATACATTTGATGAAATATTTGGTGCAAGAATCCACAGCCATGGTAACTTAGCTATTACACCCGGTTTGAAAGATCGTTATTACGTGCTCCATGTTTTCAAACGCGTATTGTCTGTGGAGAAAGACACTTTGATGGTGGATCTTGAAGGATCGACCATGATATCTAATAGGCGTTATAACTGTTGGGCCTCGTTTAACGATCTTGAACATGACTCATGTAACGGTGTTTATGCGAATATAAGCAAAAATGCCATTTTAGTTTACTATTGTTGGATGTCCGATGCTATGTCTAAGGCATCGACCTTTGTATCTTACGATCTTGATTATTTAGGTTAATCCTGAATAAAATGGCGTTAAGATTGAATATATACATAAAAAATAAAAACAAGCAATTTTATTGCAATGACTTTGGTTGTGTAGGATTACAATTATTGTTAATACATTCCTGGACCGTCGTCCTAACTAGCTCGCTTAATTGGGCCACTGACATCGTTATGGTTGATTGGGCCCTCTGTAATCCAGCTTGTGATGCTGAATCCCCGGGGTCTAATGCGCTAGTTCCTAGCTGGTTGAGTTCTCTATATGGATGTAGCGCATTCTCCACTTCTGATTCAGTTTGTGGGTTTGTAAGCCCAATAGTGCTCCTTGAAGCCCATGAATCACCTGGTTGTAATTCAATTGGGCCTGTTAGTCCAATTCTTGACAATGATCTGGACCTCAACGTTTTCCTCTCCCATTTTCCGTAGTCCACATGGGAGAAATCAACATCTCTATGTGAAAATTGTTTGGAGTGAATTTTCACTGTTGGTGCCCGGAAGGGTATATCCACTGAATGTTTAGCTGTTGACAATTTCAATTTCCCTTTAAACTTGGCAAAATGTGTTCGTTGATGCACATTTGTGTCGCTAACCCTGTAATAGAGTTTCCACGGAATAGGGTCTTTTAGCGAGAAGAATGAAGATGAAAAATAGTGGAGATCTATGTTGCATCTTAAGGGAAATGTCCAAGACGCCTGTAAGGATTCATCGTCAGTCATCCTCTTATCATGGATCTCTACAATCACCGTCCCGGTTGCGTTAATTGGTACTTGCTGTCTATATTCGATGACGCAATGGTCGATTTTCATACAGCTACGATTAAGCCTTGCTGTAAATTGCGCTGCAGTTGAAGGGAATTGCAGTACTATTTCAGTAAGATCATGACATAGCTGATATTCGTCTCTATGAGACTCTATGTAATTAAACGCATTTGGAGCATTTGCTAGCTGAGAATCCATCTATAGAAATCTGCCCGCGCAGCGGCAGAGGCTTCAATAATATCTGGGTTCAAGAATAATTGTGTTGAAGAACACAACAATGAAACGACTTGTTTCACAAGGAGGAAGAGAGAATATCTGGGCTTGTTAACAAGAAGAAAATAATGATATTTTAAAGGAATGATGAATATGTTATGTTATTCCGATAACATGAGTATAATAATCTAAGTCCACCTTGTTTATATAGAGAGTGATTTTCCAGAAGTTGGAAACGCCTTGAATTTCATTTTAGTGGCATTATTGTAAATATGGGTGTTCCCCCGATAGCGCTCTCGCTCAAAACTCCTATGAATTGGGGGAACTGGGGGAACTTATATAGTAGAAGTTCCTAAAGGCAAATCAACACGTGGCGGCCATCCGTTATAGTATT',
          }),
        ]),
      }),
    ]),
    'legacy_id': None,
    'name': 'Cabbage leaf curl Jamaica virus',
    'schema': dict({
      'molecule': dict({
        'strandedness': 'double',
        'topology': 'circular',
        'type': 'DNA',
      }),
      'multipartite': True,
      'segments': list([
        dict({
          'length': 2575,
          'name': 'DNA-A',
          'required': True,
        }),
        dict({
          'length': 2494,
          'name': 'DNA-B',
          'required': True,
        }),
      ]),
    }),
    'taxid': 345184,
  })
# ---
# name: TestCreateOTUCommands.test_autofill
  dict({
    'acronym': '',
    'excluded_accessions': list([
    ]),
    'legacy_id': None,
    'name': 'Cabbage leaf curl Jamaica virus',
    'schema': dict({
      'molecule': dict({
        'strandedness': 'double',
        'topology': 'circular',
        'type': 'DNA',
      }),
      'multipartite': True,
      'segments': list([
        dict({
          'length': 2575,
          'name': 'DNA-A',
          'required': True,
        }),
        dict({
          'length': 2494,
          'name': 'DNA-B',
          'required': True,
        }),
      ]),
    }),
    'taxid': 345184,
  })
# ---
# name: TestCreateOTUCommands.test_autofill_with_schema
  OTUSchema(molecule=Molecule(strandedness='double', type='DNA', topology='circular'), segments=[Segment(name='DNA-A', required=True, length=2575), Segment(name='DNA-B', required=True, length=2494)], multipartite=True)
# ---
# name: TestCreateOTUCommands.test_autoschema[1278205-accessions0]
  dict({
    'acronym': '',
    'excluded_accessions': list([
    ]),
    'legacy_id': None,
    'name': 'Dahlia latent viroid',
    'schema': dict({
      'molecule': dict({
        'strandedness': 'single',
        'topology': 'linear',
        'type': 'RNA',
      }),
      'multipartite': False,
      'segments': list([
        dict({
          'length': 342,
          'name': 'Dahlia latent viroid',
          'required': True,
        }),
      ]),
    }),
    'taxid': 1278205,
  })
# ---
# name: TestCreateOTUCommands.test_autoschema[345184-accessions1]
  dict({
    'acronym': '',
    'excluded_accessions': list([
    ]),
    'legacy_id': None,
    'name': 'Cabbage leaf curl Jamaica virus',
    'schema': dict({
      'molecule': dict({
        'strandedness': 'double',
        'topology': 'circular',
        'type': 'DNA',
      }),
      'multipartite': True,
      'segments': list([
        dict({
          'length': 2575,
          'name': 'DNA-A',
          'required': True,
        }),
        dict({
          'length': 2494,
          'name': 'DNA-B',
          'required': True,
        }),
      ]),
    }),
    'taxid': 345184,
  })
# ---
# name: TestUpdateOTU.test_success_no_exclusions
  list([
    dict({
      'acronym': '',
      'excluded_accessions': list([
      ]),
      'legacy_id': None,
      'name': 'Cabbage leaf curl Jamaica virus',
      'schema': None,
      'taxid': 345184,
    }),
  ])
# ---
